impl/sql/sequenced/sequenced.go

```diff
@@ -0,0 +1,182 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package sequenced stores a list of objects that have been sequenced.
+package sequenced
+
+import (
+	"bytes"
+	"database/sql"
+	"encoding/gob"
+	"errors"
+	"fmt"
+
+	"github.com/google/keytransparency/core/sequenced"
+	"github.com/google/keytransparency/core/transaction"
+)
+
+const (
+	insertMapRowExpr = `INSERT INTO Maps (MapID) VALUES (?);`
+	countMapRowExpr  = `SELECT COUNT(*) AS count FROM Maps WHERE MapID = ?;`
+	insertExpr       = `
+	INSERT INTO Sequenced (MapID, Epoch, Data)
+	VALUES (?, ?, ?);`
+	readExpr = `
+	SELECT Data FROM Sequenced
+	WHERE MapID = ? AND Epoch = ?;`
+	latestExpr = `
+	SELECT Epoch, Data FROM Sequenced
+	WHERE MapID = ? 
+	ORDER BY Epoch DESC LIMIT 1;`
+)
+
+var (
+	createStmt = []string{
+		`
+	CREATE TABLE IF NOT EXISTS Maps (
+		MapID   BIGINT NOT NULL,
+		PRIMARY KEY(MapID)
+	);`,
+		`
+	CREATE TABLE IF NOT EXISTS Sequenced (
+		MapID   BIGINT      NOT NULL,
+		Epoch   BIGINT      NOT NULL,
+		Data    BLOB(1024)  NOT NULL,
+		PRIMARY KEY(MapID, Epoch),
+		FOREIGN KEY(MapID) REFERENCES Maps(MapID) ON DELETE CASCADE
+	);`,
+	}
+	// ErrNotSupported occurs when performing an operaion that has been disabled.
+	ErrNotSupported = errors.New("operation not supported")
+)
+
+// Sequenced stores objects in a table.
+type Sequenced struct {
+	db *sql.DB
+}
+
+// New returns an object that can store sequenced items for multiple maps.
+func New(db *sql.DB) (sequenced.Sequenced, error) {
+	if err := db.Ping(); err != nil {
+		return nil, fmt.Errorf("No DB connection: %v", err)
+	}
+
+	if err := create(db); err != nil {
+		return nil, err
+	}
+	return &Sequenced{
+		db: db,
+	}, nil
+}
+
+// Create creates a new database.
+func create(db *sql.DB) error {
+	for _, stmt := range createStmt {
+		_, err := db.Exec(stmt)
+		if err != nil {
+			return fmt.Errorf("Failed to create appender tables: %v", err)
+		}
+	}
+	return nil
+}
+
+func (s *Sequenced) insertMapRow(txn transaction.Txn, mapID int64) error {
+	// Check if a map row does not exist for the same MapID.
+	countStmt, err := txn.Prepare(countMapRowExpr)
+	if err != nil {
+		return fmt.Errorf("insertMapRow(): %v", err)
+	}
+	defer countStmt.Close()
+	var count int
+	if err := countStmt.QueryRow(mapID).Scan(&count); err != nil {
+		return fmt.Errorf("insertMapRow(): %v", err)
+	}
+	if count >= 1 {
+		return nil
+	}
+
+	// Insert a map row if it does not exist already.
+	insertStmt, err := txn.Prepare(insertMapRowExpr)
+	if err != nil {
+		return fmt.Errorf("insertMapRow(): %v", err)
+	}
+	defer insertStmt.Close()
+	_, err = insertStmt.Exec(mapID)
+	if err != nil {
+		return fmt.Errorf("insertMapRow(): %v", err)
+	}
+	return nil
+}
+
+// Append adds an object to the append-only data structure.
+func (s *Sequenced) Write(txn transaction.Txn, mapID, epoch int64, obj interface{}) error {
+	if err := s.insertMapRow(txn, mapID); err != nil {
```

Comment:
Can `insertMapRow` be called in `New` once instead of doing that for every write? Does this support multi-homed maps? If that's the case we need to change this in all other places while we're at it. I can do that after this PR is merged.